src/dali_backend.cc

```diff
@@ -234,70 +232,119 @@ class DaliModelInstance : public ::triton::backend::BackendModelInstance {
+        } else {
+          ENFORCE(meta.shape.num_samples() == reqs_batch_sizes[ri],
+                  "Each input in a request must have the same batch size.");
+        }
       }
-      ret.push_back({input.Meta(), std::move(buffers)});
     }
-    return ret;
+    for (const auto& descrs : input_map) {
+      inputs.push_back(descrs.second);
+    }
+    return {inputs, reqs_batch_sizes};
   }
 
   int32_t GetDaliDeviceId() {
     return !CudaStream() ? ::dali::CPU_ONLY_DEVICE_ID : device_id_;
   }
 
   /**
-   * @brief Allocate outputs required by a given request.
+   * @brief Allocate outputs expected by given \p requests.
    *
-   * Lifetime of the created buffer is bound to the \p response
+   * Lifetime of the created buffer is bound to each of the \p responses
+   * @param batch_sizes batch size of each request
    */
-  std::vector<ODescr> AllocateOutputs(TritonRequestView request, TritonResponseView response,
+  std::vector<ODescr> AllocateOutputs(const std::vector<TritonRequest>& requests,
+                                      const std::vector<TritonResponse>& responses,
+                                      const std::vector<int>& batch_sizes,
                                       const std::vector<OutputInfo>& outputs_info) {
-    uint32_t output_cnt = request.OutputCount();
+    assert(requests.size() > 0);
+    assert(requests.size() == responses.size());
+    assert(requests.size() == batch_sizes.size());
+    uint32_t output_cnt = requests[0].OutputCount();
+    for (auto& req : requests) {
+      ENFORCE(output_cnt == req.OutputCount(),
+              "All of the requests must expect the same number of outputs.");
+    }
     ENFORCE(outputs_info.size() == output_cnt,
-            make_string("Number of outputs in the model configuration (", output_cnt,
-                        ") does not match to the number of outputs from DALI pipeline (",
-                        outputs_info.size(), ")"));
+            make_string("Number of outputs expected by the requests (", output_cnt,
+                        ") does not match the number of outputs from DALI pipeline (",
+                        outputs_info.size(), ")."));
     const auto& output_indices = dali_model_->GetOutputOrder();
+    ENFORCE(output_cnt == output_indices.size(),
+            make_string("Number of outputs exptected by the requests (", output_cnt,
+                        ") does not match the number of outputs in the config (",
+                        output_indices.size(), ")."));
     std::vector<ODescr> outputs(output_cnt);
     outputs.reserve(output_cnt);
-    for (uint32_t i = 0; i < output_cnt; ++i) {
-      auto name = request.OutputName(i);
-      int output_idx = output_indices.at(name);
+    for (const auto& out_index : output_indices) {
+      auto name = out_index.first;
+      int output_idx = out_index.second;
+      auto shapes = split_list_shape(outputs_info[output_idx].shape, batch_sizes);
```

Comment:
I assume that we execute those merged batches with DALI, that means we need to translate/slice it back for triton output?

Reply:
Yes. DaliExecutor handles 'chunked outputs', meaning the output can be copied to non-contiguous buffers.